tests/baselines/reference/discriminateObjectTypesOnly.errors.txt

```diff
@@ -1,17 +0,0 @@
-tests/cases/compiler/discriminateObjectTypesOnly.ts(9,7): error TS2741: Property 'toFixed' is missing in type '{ toString: undefined; }' but required in type '{ toFixed: null; toString: undefined; }'.
-
-
-==== tests/cases/compiler/discriminateObjectTypesOnly.ts (1 errors) ====
-    type Thing = number | object;
-    const k: Thing = { toFixed: null }; // OK, satisfies object
-    
-    type Thing2 = number | { toFixed: null } | object;
-    const q: Thing2 = { toFixed: null };
-    const h: Thing2 = { toString: null }; // OK, satisfies object
-    
-    type Thing3 = number | { toFixed: null, toString: undefined } | object;
-    const l: Thing3 = { toString: undefined }; // error, toFixed isn't null
-          ~
-!!! error TS2741: Property 'toFixed' is missing in type '{ toString: undefined; }' but required in type '{ toFixed: null; toString: undefined; }'.

```

Comment:
IMO, this should still be an error - #31708 came up because we unconditionally treated `object` as a type with no properties for excess property checking (so all usages of `object` used to trigger excess property errors like this), which caused problems in these primitive only unions (where there really isn't any hinting as to what properties an object literal should contain, really), however in case like this one, where there's an explicit object type, using the hints from the specified object types for excess properties is more desirable (and prevents you form passing total garbage and typo'ing the property names from the provided literal).

Reply:
This PR doesn't change anything with respect to #31708. That logic is still in place and we still filter out primitives from unions containing `object` before we do excess property checking. Also, there was never any logic in place to remove `object` when actual object types are present. Rather, when `object` is present in a union it effectively turns off excess property checking, just like `{}`. However, we had some very odd behavior when it came to using `toString` and the other property names from `Object`. For example:

```ts
let x1: { a: string, b: number } | object = { a: 'test' };
let x2: { a: 'test', b: number } | object = { a: 'test' };
let x3: { toString: string, b: number } | object = { toString: 'test' };
let x4: { toString: 'test', b: number } | object = { toString: 'test' };  // Error!
```

Above, we would error on `x4`, but not on any of the other ones. That just doesn't make sense, and we now permit it. (Keep in mind here that by design we don't give special treatment to `toString` and the other `Object` properties in assignment checking.)

Reply:
As I said in person - `{}` disables excess property checking because it has baggage whereupon it was used as a general top type. `object` has no such compunctions. Otherwise there's no way to _have_ an excess property checked empty object.

Reply:
What I'm pointing out is that, except for the logic that filters out primitives, `object` and `{}` behave the same for purposes of excess property checking in the current compiler. It's existing behavior and unrelated to the issue I'm fixing in this PR.